design-documents/service-isolation/inter-service-communication.md

```diff
@@ -0,0 +1,294 @@
+    ],
+];
+```
+The invoker will look up the information based on provided services' class names.
+ 
+#### Changes to RESTful Web API mechanism
+We need to able so send and receive more information when using RESTful web API for inter-service communication than for regular
+communications so to distinguish regular endpoints and endpoints meant for inter-service communications we will declare them in a
+seperate yet similar xml files - _inter_webapi.xml_. Since we declare services by declaring PHP interfaces there is no reason to provide
+human-readable URLs for endpoints meant for inter-service communications, thus we'll make it possible to omit _url_ attribute of the
+_route_ element in inter_webapi.xml and have it autogenerated for us. This way _the invoker_ will be able to determine URL
+for a service based on class name/method name provided. The _method_ attribute will be dropped because otherwise both nodes
+who are calling remote services and nodes hosting the services will have to have the inter_webapi.xml installed.
+Also caching service call results via
+web server (like for services accessed by GET) may be dangerous because the most fresh results are expected from service calls.
+POST method will always be used when sending HTTP requests to remote services.
+_resources_ child element of the _route_ element will be dropped as well because when calling a service remotely
+we don't need to check privileges again - requests are initiated by our code so we assume they're just. In the end enabling a service
+to be called by other remote services will look like this:
+_inter_webapi.xml_
+```xml
+<?xml version="1.0"?>
+<services xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Webapi:etc/inter_webapi.xsd">
+
+    <!-- Product Service -->
+    <service class="Magento\Catalog\Api\ProductRepositoryInterface" method="save"/>
+</services>
+```
+ 
+We will have to pass authenticated user information to remote services so we'll add _X-Magento-User-Type_ and _X-Magento-User-Id_
+headers to requests for a _UserContextInterface_ instance to be initialized. There's no need for nodes to validate auth information of
+incoming
+requests since those requests will come from other nodes in a closed network or they will be encrypted. These headers will only be
+accepted by endpoints declared in _inter_webapi.xml_ files.
+ 
+The way exceptions thrown by services are serialized right now for RESTful web API needs to be changed for inter-service endpoints.
+Exceptions are serialized without their type information provided - this info is crutial for services to communicate with each other so
+the type information will be added to the output.
+ 
+_inter_webapi.xml_ files will be contained in separate <ModuleName>InterApi modules and will have to be installed
+on nodes accepting remote inter-service requests.
+ 
+##### Changes for retries mechanism
```

Comment:
I think we shouldn't try to implement this functionality ourselves. What are the reasons to not use service mesh?

Reply:
What is the purpose of default implementation? Development?